ethapi/api_test.go

```diff
@@ -399,3 +401,235 @@ func setExpectedStateCalls(mockState *state.MockStateDB) {
+			Gas:      1e6,
+			GasPrice: big.NewInt(500e9),
+			AccessList: types.AccessList{
+				{Address: common.Address{1}, StorageKeys: []common.Hash{{0x01}}},
+			},
+		},
+		"dynamicFee": &types.DynamicFeeTx{
+			Nonce:     2,
+			To:        &common.Address{1},
+			Gas:       1e6,
+			GasFeeCap: big.NewInt(500e9),
+			GasTipCap: big.NewInt(500e9),
+		},
+		"blob empty list": &types.BlobTx{
+			Nonce:      3,
+			Gas:        1e6,
+			GasFeeCap:  uint256.NewInt(500e9),
+			BlobFeeCap: uint256.NewInt(500e9),
+		},
+		"blob": &types.BlobTx{
+			Nonce:      3,
+			Gas:        1e6,
+			GasFeeCap:  uint256.NewInt(500e9),
+			BlobFeeCap: uint256.NewInt(500e9),
+			BlobHashes: []common.Hash{{0x01}},
+		},
+		"setCode empty list": &types.SetCodeTx{
+			Nonce: 4,
+			To:    common.Address{42},
+			Gas:   1e6,
+		},
+		"setCode": &types.SetCodeTx{
+			Nonce: 4,
+			To:    common.Address{42},
+			Gas:   1e6,
+			AuthList: []types.SetCodeAuthorization{
+				authorization,
+			},
+		},
+	}
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			signed := signTransaction(t, big.NewInt(1), test, key)
+
+			blockHash := common.Hash{1, 2, 3, 4}
+			blockNumber := uint64(4321)
+			index := uint64(0)
+			baseFee := big.NewInt(1234)
+
+			rpcTx := newRPCTransaction(signed, blockHash, blockNumber, index, baseFee)
+			require.Equal(t, signed.Hash(), rpcTransactionToTransaction(t, rpcTx).Hash())
+
+			encoded, err := json.Marshal(rpcTx)
+			require.NoError(t, err)
+
+			decoded := new(RPCTransaction)
+			err = json.Unmarshal(encoded, decoded)
+			require.NoError(t, err)
+			require.Equal(t, blockHash, *decoded.BlockHash)
+			require.Equal(t, int64(blockNumber), decoded.BlockNumber.ToInt().Int64())
+			require.Equal(t, index, uint64(*decoded.TransactionIndex))
+			require.Equal(t, signed.Hash(), rpcTransactionToTransaction(t, decoded).Hash())
+		})
+	}
+}
+
+func signTransaction(
+	t *testing.T,
+	chainId *big.Int,
+	payload any,
+	from *ecdsa.PrivateKey,
+) *types.Transaction {
+	t.Helper()
+
+	switch tx := payload.(type) {
+	case *types.LegacyTx:
+		res, err := types.SignTx(
+			types.NewTx(tx),
+			types.NewEIP155Signer(chainId),
+			from)
+		require.NoError(t, err)
+		return res
+	case *types.AccessListTx:
+		res, err := types.SignTx(
+			types.NewTx(tx),
+			types.NewEIP2930Signer(chainId),
+			from)
+		require.NoError(t, err)
+		return res
+	case *types.DynamicFeeTx:
+		res, err := types.SignTx(
+			types.NewTx(tx),
+			types.NewLondonSigner(chainId),
+			from)
+		require.NoError(t, err)
+		return res
+	case *types.BlobTx:
+		res, err := types.SignTx(
+			types.NewTx(tx),
+			types.NewCancunSigner(chainId),
+			from)
+		require.NoError(t, err)
+		return res
+	case *types.SetCodeTx:
+		res, err := types.SignTx(
+			types.NewTx(tx),
+			types.NewPragueSigner(chainId),
+			from)
+		require.NoError(t, err)
+		return res
+	default:
+		t.Error("unsupported transaction type ", reflect.TypeOf(payload))
+		return nil
+	}
+}
```

Comment:
cosmtic suggestion, wdyt ? 
```suggestion
var signer types.Signer
	var txTyped *types.Transaction
	switch tx := payload.(type) {
	case *types.LegacyTx:
		txTyped = types.NewTx(tx)
		signer = types.NewEIP155Signer(chainId)
	case *types.AccessListTx:
		txTyped = types.NewTx(tx)
		signer = types.NewEIP2930Signer(chainId)
	case *types.DynamicFeeTx:
		txTyped = types.NewTx(tx)
		signer = types.NewLondonSigner(chainId)
	case *types.BlobTx:
		txTyped = types.NewTx(tx)
		signer = types.NewCancunSigner(chainId)
	case *types.SetCodeTx:
		txTyped = types.NewTx(tx)
		signer = types.NewPragueSigner(chainId)
	default:
		t.Error("unsupported transaction type ", reflect.TypeOf(payload))
		return nil
	}
	res, err := types.SignTx(
		txTyped,
		signer,
		from)
	require.NoError(t, err)
	return res
}
```

Reply:
Here you can see the influence of functional programming. I developed a resistance to use mutable variables, specially when they are default initialized.
On the other hand, your code is smaller and maybe cleaner (the possibility of the nil txTyped after the switch still freaks me out)

Reply:
right. I did not consider the nil case the `default` case does an early return. You could also add a `require.NotEmpty(txType)` for safety

Reply:
The code was simplified quite a lot. The latest available signer is capable of signing all the previous transaction types (with the exception of very old ones before eip-155, which do not support)

Reply:
https://github.com/0xsoniclabs/sonic/pull/81/commits/e6f79b84bbeaeb796efbdf5df14001111e5c2948